src/marin/post_training/rollout_worker.py

```diff
@@ -0,0 +1,332 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Inference worker for RL/post-training rollout generation.
+
+This worker loads model checkpoints, generates rollouts from a single environment,
+and writes the rollout data to files for training workers to consume.
+"""
+
+import asyncio
+import logging
+import os
+import socket
+import threading
+import time
+from dataclasses import dataclass
+from typing import Any
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+import requests
+from levanter.inference.openai import InferenceServer, InferenceServerConfig
+
+from marin.post_training.environments.marin_env import InferenceContext
+
+from .flax.utils import (
+    jax_distributed_barrier,
+)
+from .rl_dataset import create_dataset_from_environment
+from .rollout_storage import RolloutBatch, RolloutWriter, TaggedRolloutBatch
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class InferenceWorkerConfig:
+    """Configuration for InferenceWorker."""
+
+    inference_server_config: InferenceServerConfig
+    policy_model: Any  # We'll type this properly later
+    reference_model: Any  # We'll type this properly later
+    environment_spec: str
+    rollout_writer: RolloutWriter
+    environment: Any
+    environment_name: str
+    max_input_length: int
+    max_output_length: int
+    pad_token_id: int
+    n_prompts_per_step: int
+    n_generations: int
+    temperature: float
+    log_freq: int
+    rollout_batch_size: int = 32
+    max_rollouts: int | None = None
+
+
+class LevanterInferenceContext(InferenceContext):
+    """Context that uses Levanter model and inference server."""
+
+    model: Any
+    inference_server: InferenceServer
+    max_tokens: int
+    _tokenizer: Any
+
+    def __init__(self, model, inference_server: InferenceServer, max_tokens: int):
+        self.model = model
+        self.inference_server = inference_server
+        self.max_tokens = max_tokens
+        self._tokenizer = inference_server.inference_context.tokenizer
+
+    @property
+    def tokenizer(self):
+        return self._tokenizer
+
+    def generate(
+        self,
+        prompts: list[str],
+        temperature: float = 1.0,
+        n_generations: int = 1,
+        top_p: float | None = None,
+        top_k: int | None = None,
+    ) -> list[list[dict]]:
+        """Generate responses for a batch of prompts."""
+        host = self.inference_server.config.host
+        port = self.inference_server.config.port
+        url = f"http://{host}:{port}/v1/chat/completions"
```

Comment:
too good for the OAI library?

Reply:
Switched over to OAI & added parallelism.